apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.StringParsing.cs

```diff
@@ -0,0 +1,191 @@
+using System.Linq;
+using System.Text;
+using Google.Cloud.Spanner.V1;
+using TypeCode = Google.Cloud.Spanner.V1.TypeCode;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Represents a Type that can be stored in a Spanner column or returned from a query.
+    /// </summary>
+    public sealed partial class SpannerDbType
+    { 
+        /// <summary>
+        /// Given a string representation, returns an instance of <see cref="SpannerDbType"/>.
+        /// </summary>
+        /// <param name="spannerType">A string representation of a SpannerDbType. See <see cref="ToString"/></param>
+        /// <param name="spannerDbType">If parsing was successful, then an instance of <see cref="SpannerDbType"/>.
+        ///  Otherwise null.</param>
+        /// <returns></returns>
+        public static bool TryParse(string spannerType, out SpannerDbType spannerDbType)
+        {
+            spannerDbType = null;
+            if (!TryParsePartial(spannerType, out TypeCode code, out int? size, out string remainder))
+            {
+                return false;
+            }
+            switch (code)
+            {
+                case TypeCode.Unspecified:
+                case TypeCode.Bool:
+                case TypeCode.Int64:
+                case TypeCode.Float64:
+                case TypeCode.Timestamp:
+                case TypeCode.Date:
+                case TypeCode.String:
+                case TypeCode.Bytes:
+                    if (!string.IsNullOrEmpty(remainder))
+                    {
+                        //unexepected inner remainder on simple type
+                        return false;
+                    }
+                    spannerDbType = !size.HasValue ? FromTypeCode(code) : new SpannerDbType(code, size.Value);
+                    return true;
+                case TypeCode.Array:
+                    if (!TryParse(remainder, out SpannerDbType elementType))
+                    {
+                        return false;
+                    }
+                    spannerDbType = new SpannerDbType(code, elementType);
+                    return true;
+                case TypeCode.Struct:
+                    //there could be nested structs, so we need to be careful about parsing the inner string.
+                    List<Tuple<string, SpannerDbType>> fields = new List<Tuple<string, SpannerDbType>>();
+                    int currentIndex = 0;
+                    while (currentIndex < remainder.Length)
+                    {
+                        int midfieldIndex = NonNestedIndexOf(remainder, currentIndex, ':');
+                        if (midfieldIndex == -1)
+                        {
+                            return false;
+                        }
+                        int endFieldIndex = NonNestedIndexOf(remainder, currentIndex, ',');
+                        if (endFieldIndex == -1)
+                        {
+                            //we reached the last field.
```

Comment:
a string with unbalanced brackets or parens will fail when processing the inner remainder.

For example
STRUCT<A:STRUCT<B:STRING>
will recurse as:
code = STRUCT, remainder = A:STRUCT<B:STRING
FAIL A:STRUCT<B:STRING has no ending bracket (line 159 below)

I'm sure there are more cases here worth testing though.  I need to add negative tests.

Reply:
another way to think about it is that given the string "A:B", if B ends up including more stuff than it should for any reason right here, B would fail to parse as a valid type itself.
I should also be validating that A is a valid field name (no spaces, punctuation, etc).